docs/config.md
# Configuration Reference

```yaml
gos_web_socket:
  authentication:
    providers:
      session:

        # The service ID of the session handler service used to read session data.
        session_handler:      null

        # The firewalls from which the session token can be used; can be an array, a string, or null to allow all firewalls.
        firewalls:            null
    storage:

      # The type of storage for the websocket server authentication tokens.
      type:                 in_memory # One of "in_memory"; "psr_cache"; "service", Required

      # The cache pool to use when using the PSR cache storage.
      pool:                 null

      # The service ID to use when using the service storage.
      id:                   null
  server:

    # The host IP address on the server which connections for the websocket server are accepted.
    host:                 ~ # Required

    # The port on the server which connections for the websocket server are accepted.
    port:                 ~ # Required

    tls:

      # Enables the native tls support that can be configured with the options below.
      enabled: false

      # The options to set up the tls configuration. See the example below or see https://www.php.net/manual/en/context.ssl.php for all available options.
      options:
        local_cert: '/path/to/cert/cert.crt'
        local_pk: '/path/to/key/mykey.key'
        verify_peer: false

    # Enables checking the Origin header of websocket connections for allowed values.
    origin_check:         false

    # Enables checking the originating IP address of websocket connections for blocked addresses.
    ip_address_check:     false

    # Flag indicating a keepalive ping should be enabled on the server.
    keepalive_ping:       false

    # The time in seconds between each keepalive ping.
    keepalive_interval:   30
    router:
      resources:

        # Prototype
        -
          resource:             ~ # Required
          type:                 null # One of "closure"; "container"; "glob"; "php"; "xml"; "yaml"; null
  ping:
    services:

      # Prototype
      -

        # The name of the service to ping.
        name:                 ~ # Required

        # The type of the service to be pinged.
        type:                 ~ # One of "doctrine"; "pdo", Required

        # The time (in seconds) between executions of this ping.
        interval:             20

  # A list of origins allowed to connect to the websocket server, must match the value from the "Origin" header of the HTTP request.
  origins:              []

  # A list of IP addresses which are not allowed to connect to the websocket server.
  blocked_ip_addresses: []
```


docs/ConfigurationReference.md
# Configuration Reference

```yaml
gos_web_socket:
  client:

    # The service ID of the session handler service used to read session data.
    session_handler:      ~

    # The name of the security firewall to load the authenticated user data for.
    firewall:             ws_firewall
    storage:

      # The service ID of the storage driver to use for storing connection data.
      driver:               gos_web_socket.client.driver.in_memory

      # The cache TTL (in seconds) for clients in storage.
      ttl:                  900

      # The service ID of a decorator for the client storage driver.
      decorator:            ~
  shared_config:        true
  server:

    # The host IP address on the server which connections for the websocket server are accepted.
    host:                 ~ # Required

    # The port on the server which connections for the websocket server are accepted.
    port:                 ~ # Required

    # Enables checking the Origin header of websocket connections for allowed values.
    origin_check:         false

    # Enables checking the originating IP address of websocket connections for blocked addresses.
    ip_address_check:     false

    # Flag indicating a keepalive ping should be enabled on the server.
    keepalive_ping:       false

    # The time in seconds between each keepalive ping.
    keepalive_interval:   30
    router:
      resources:

        # Prototype
        -
          resource:             ~ # Required
          type:                 null # One of "closure"; "container"; "glob"; "php"; "xml"; "yaml"; null

  # A list of origins allowed to connect to the websocket server, must match the value from the "Origin" header of the HTTP request.
  origins:              []

  # A list of IP addresses which are not allowed to connect to the websocket server.
  blocked_ip_addresses: []
  ping:
    services:

      # Prototype
      -

        # The name of the service to ping.
        name:                 ~ # Required

        # The type of the service to be pinged.
        type:                 ~ # One of "doctrine"; "pdo", Required
```


CHANGELOG.md
# Changelog

## 3.12.0 (2021-11-03)

- [#454](https://github.com/GeniusesOfSymfony/WebSocketBundle/pull/454) Add native TLS support
- Add support for command auto-completion for Symfony 5.4
- Add attributes to silence PHP 8.1 deprecations

## 3.11.0 (2021-08-29)

- Implemented a new [authentication API](docs/authentication.md) and deprecated the previous implementation

## 3.10.0 (2021-07-31)

- Add support for `psr/log` 2.0 and 3.0
- Updated use of `react/event-loop` to prefer newer API
- Drop support for Symfony 5.2 (Branch is EOL)

## 3.9.0 (2021-06-30)

- Add support for configuring the ping interval for ping services
- Deprecate public access to public container services
- Deprecated the `setTimeout` method of periodic classes
- Deprecated the `getTimeout` method of `Gos\Bundle\WebSocketBundle\Periodic\PeriodicInterface` in favor of `getInterval`
- Deprecated the `shared_config` config node, in 4.0 the Twig globals will automatically be created when Twig is available

## 3.8.1 (2021-06-16)

- Fixes for IP blocker service

## 3.8.0 (2021-06-16)

- Add support for `doctrine/cache` 2.0
- Bump minimum supported `doctrine/dbal` versions
- Add the ability to block IP addresses from connecting to the websocket server
- Deprecated the `Gos\Bundle\WebSocketBundle\Event\ClientRejectedEvent` class and corresponding event, subscribe to `Gos\Bundle\WebSocketBundle\Event\ConnectionRejectedEvent` instead

## 3.7.2 (2021-04-25)

- Symfony 5.3 compatibility fixes

## 3.7.1 (2021-03-29)

- Fix incorrect service reference

## 3.7.0 (2021-03-28)

- Drop support for Symfony 5.1 (Branch is EOL)
- Deprecated the `WS` class and `Socket` global variable in the `websocket.js` file, use the new static `GosSocket.connect()` method as a replacement for `WS.connect()` and store the singleton within your application if necessary
- Deprecated the `Gos\Bundle\WebSocketBundle\Server\EntryPoint` class, use the `Gos\Bundle\WebSocketBundle\Server\ServerLauncher` class instead
- Deprecated `Gos\Bundle\WebSocketBundle\Server\App\PushableWampServerInterface`, `Gos\Bundle\WebSocketBundle\Topic\PushableTopicInterface`,`Gos\Bundle\WebSocketBundle\Server\App\Dispatcher\TopicDispatcherInterface::onPush()`, and `Gos\Bundle\WebSocketBundle\Server\App\WampApplication::onPush()`; this is a continuation of the deprecation of the pusher services which was not previously flagged
- Deprecated `Gos\Bundle\WebSocketBundle\Topic\TopicManager::setWampApplication()`, as of 4.0 the dependency will be injected through the class' constructor

## 3.6.2 (2021-01-18)

- Reverted the change in 3.6.1

## 3.6.1 (2021-01-12)

- [#422](https://github.com/GeniusesOfSymfony/WebSocketBundle/pull/422) Periodic timer becomes broken when all topic subscribers closed their connections

## 3.6.0 (2020-11-27)

- Deprecated the requirement to implement `Gos\Bundle\WebSocketBundle\Server\App\Dispatcher\TopicDispatcherInterface::dispatch()`, the method will be removed from the interface in 4.0
- [MINOR B/C BREAK] Added type checks to the `$topic` argument of `Gos\Bundle\WebSocketBundle\Topic\TopicManager::getTopic()`, an `InvalidArgumentException` will now be thrown if a non-supported type is provided

## 3.5.0 (2020-11-02)

- Drop support for Symfony 5.0 (Branch is EOL)
- Allow install on PHP 8
- Updates for changes introduced in `doctrine/dbal` 2.11

## 3.4.0 (2020-08-26)

- Deprecated support for the `gos/websocket-client` package, use `ratchet/pawl` instead
- Deprecated the `Gos\Bundle\WebSocketBundle\Client\Driver\DoctrineCacheDriverDecorator`, if using the `doctrine/cache` package a `Gos\Bundle\WebSocketBundle\Client\Driver\SymfonyCacheDriverDecorator` using a `Symfony\Component\Cache\DoctrineProvider` instance can be used
- Introduced `Gos\Bundle\WebSocketBundle\Client\ClientStorageInterface::removeAllClients()`, if implemented a client store can remove all clients from storage; in 4.0, this method will be required on all interface implementations
- Introduced `Gos\Bundle\WebSocketBundle\Client\Driver\DriverInterface::clear()`, if implemented a driver can remove all data from storage; in 4.0, this method will be required on all interface implementations
- `Gos\Bundle\WebSocketBundle\EventListener\StartServerListener` will now support `SIGTERM` signals as well as `SIGINT`

## 3.3.0 (2020-07-06)

- Change `Gos\Bundle\WebSocketBundle\Periodic\DoctrinePeriodicPing` to address the deprecation of `Doctrine\DBAL\Driver\PingableConnection`
- Deprecate support for `Doctrine\DBAL\Driver\PingableConnection` implementations in `Gos\Bundle\WebSocketBundle\Periodic\DoctrinePeriodicPing`, in 4.0 `Doctrine\DBAL\Connection` instances will be required
- Add aliases to bundle events to allow registering listeners using the FQCN
- Deprecated `Gos\Bundle\WebSocketBundle\Event\ClientErrorEvent::setException()`, in 4.0 a `Throwable` instance will be a required constructor argument
- Deprecated `Gos\Bundle\WebSocketBundle\Event\ClientErrorEvent::getException()`, use `Gos\Bundle\WebSocketBundle\Event\ClientErrorEvent::getThrowable()` instead
- Remove call to `Ratchet\Wamp\Topic::broadcast()` if the dispatch method fails, see [the security advisory](https://github.com/GeniusesOfSymfony/WebSocketBundle/security/advisories/GHSA-wwgf-3xp7-cxj4) for additional details

## 3.2.0 (2020-06-01)

- Extend `Gos\Component\WebSocketClient\Wamp\ClientFactoryInterface` inside `Gos\Bundle\WebSocketBundle\Pusher\Wamp\WampConnectionFactoryInterface`
- Added new `gos_web_socket.websocket_client` configuration node to configure a `Gos\Component\WebSocketClient\Wamp\ClientInterface` instance

## 3.1.0 (2020-05-31)

- Use the `symfony/deprecation-contracts` package to trigger runtime deprecation notices
- Deprecated `Gos\Bundle\WebSocketBundle\Pusher\PusherInterface` and `Gos\Bundle\WebSocketBundle\Pusher\ServerPushHandlerInterface`, and all related services, in favor of the Symfony Messenger component

src/DependencyInjection/Configuration.php
<?php declare(strict_types=1);

namespace Gos\Bundle\WebSocketBundle\DependencyInjection;

use Gos\Bundle\WebSocketBundle\DependencyInjection\Factory\Authentication\AuthenticationProviderFactoryInterface;
use Symfony\Component\Config\Definition\Builder\ArrayNodeDefinition;
use Symfony\Component\Config\Definition\Builder\TreeBuilder;
use Symfony\Component\Config\Definition\ConfigurationInterface;

/**
 * @author Johann Saunier <johann_27@hotmail.fr>
 */
final class Configuration implements ConfigurationInterface
{
    public const PING_SERVICE_TYPE_DOCTRINE = 'doctrine';
    public const PING_SERVICE_TYPE_PDO = 'pdo';

    public const AUTHENTICATION_STORAGE_TYPE_IN_MEMORY = 'in_memory';
    public const AUTHENTICATION_STORAGE_TYPE_PSR_CACHE = 'psr_cache';
    public const AUTHENTICATION_STORAGE_TYPE_SERVICE = 'service';

    /**
     * @var AuthenticationProviderFactoryInterface[]
     */
    private array $authenticationProviderFactories;

    /**
     * @param AuthenticationProviderFactoryInterface[] $authenticationProviderFactories
     */
    public function __construct(array $authenticationProviderFactories)
    {
        $this->authenticationProviderFactories = $authenticationProviderFactories;
    }

    public function getConfigTreeBuilder(): TreeBuilder
    {
        $treeBuilder = new TreeBuilder('gos_web_socket');

        $rootNode = $treeBuilder->getRootNode();

        $this->addAuthenticationSection($rootNode);
        $this->addServerSection($rootNode);
        $this->addPingSection($rootNode);

        $rootNode->children()
            ->arrayNode('origins')
                ->info('A list of origins allowed to connect to the websocket server, must match the value from the "Origin" header of the HTTP request.')
                ->scalarPrototype()
                ->validate()
                    ->ifInArray(['localhost', '127.0.0.1'])
                        ->thenInvalid('%s is added by default')
                    ->end()
                ->end()
            ->end()
            ->arrayNode('blocked_ip_addresses')
                ->info('A list of IP addresses which are not allowed to connect to the websocket server.')
                ->scalarPrototype()
                ->end()
            ->end()
        ->end();

        return $treeBuilder;
    }

    private function addAuthenticationSection(ArrayNodeDefinition $rootNode): void
    {
        $authenticationNode = $rootNode->children()
            ->arrayNode('authentication')
                ->addDefaultsIfNotSet();

        $this->addAuthenticationProvidersSection($authenticationNode);

        $authenticationNode->children()
                ->arrayNode('storage')
                    ->addDefaultsIfNotSet()
                    ->children()
                        ->enumNode('type')
                            ->isRequired()
                            ->defaultValue(self::AUTHENTICATION_STORAGE_TYPE_IN_MEMORY)
                            ->info('The type of storage for the websocket server authentication tokens.')
                            ->values([self::AUTHENTICATION_STORAGE_TYPE_IN_MEMORY, self::AUTHENTICATION_STORAGE_TYPE_PSR_CACHE, self::AUTHENTICATION_STORAGE_TYPE_SERVICE])
                        ->end()
                        ->scalarNode('pool')
                            ->defaultNull()
                            ->info('The cache pool to use when using the PSR cache storage.')
                        ->end()
                        ->scalarNode('id')
                            ->defaultNull()
                            ->info('The service ID to use when using the service storage.')
                        ->end()
                    ->end()
                    ->validate()
                        ->ifTrue(static fn (array $config): bool => ('' === $config['pool'] || null === $config['pool']) && self::AUTHENTICATION_STORAGE_TYPE_PSR_CACHE === $config['type'])
                        ->thenInvalid('A cache pool must be set when using the PSR cache storage')
                    ->end()
                    ->validate()
                        ->ifTrue(static fn (array $config): bool => ('' === $config['id'] || null === $config['id']) && self::AUTHENTICATION_STORAGE_TYPE_SERVICE === $config['type'])
                        ->thenInvalid('A service ID must be set when using the service storage')
                    ->end()
                ->end()
            ->end()
        ->end();
    }

    private function addAuthenticationProvidersSection(ArrayNodeDefinition $authenticationNode): void
    {
        $providerNodeBuilder = $authenticationNode
            ->fixXmlConfig('provider')
            ->children()
                ->arrayNode('providers')
                    // ->requiresAtLeastOneElement() // Will be required as of 4.0
        ;

        foreach ($this->authenticationProviderFactories as $factory) {
            $name = str_replace('-', '_', $factory->getKey());
            $factoryNode = $providerNodeBuilder->children()->arrayNode($name)->canBeUnset();

            $factory->addConfiguration($factoryNode);
        }
    }

    private function addServerSection(ArrayNodeDefinition $rootNode): void
    {
        $rootNode->children()
            ->arrayNode('server')
                ->addDefaultsIfNotSet()
                ->children()
                    ->scalarNode('host')
                        ->isRequired()
                        ->cannotBeEmpty()
                        ->info('The host IP address on the server which connections for the websocket server are accepted.')
                    ->end()
                    ->scalarNode('port')
                        ->isRequired()
                        ->info('The port on the server which connections for the websocket server are accepted.')
                    ->end()
                    ->arrayNode('tls')
                        ->addDefaultsIfNotSet()
                        ->children()
                            ->booleanNode('enabled')
                                ->info('Enables native TLS support.')
                                ->defaultFalse()
                            ->end()
                            ->variableNode('options')
                                ->info('An array of options for the TLS context, see https://www.php.net/manual/en/context.ssl.php for available options.')
                                ->defaultValue([])
                            ->end()
                        ->end()
                    ->end()
                    ->booleanNode('origin_check')
                        ->defaultFalse()
                        ->info('Enables checking the Origin header of websocket connections for allowed values.')
                    ->end()
                    ->booleanNode('ip_address_check')
                        ->defaultFalse()
                        ->info('Enables checking the originating IP address of websocket connections for blocked addresses.')
                    ->end()
                    ->booleanNode('keepalive_ping')
                        ->defaultFalse()
                        ->info('Flag indicating a keepalive ping should be enabled on the server.')
                    ->end()
                    ->integerNode('keepalive_interval')
                        ->defaultValue(30)
                        ->info('The time in seconds between each keepalive ping.')
                    ->end()
                    ->arrayNode('router')
                        ->children()
                            ->arrayNode('resources')
                                ->beforeNormalization()
                                    ->ifTrue(static function ($v): bool {
                                        foreach ($v as $resource) {
                                            if (!\is_array($resource)) {
                                                return true;
                                            }
                                        }

                                        return false;
                                    })
                                    ->then(static function ($v): array {
                                        $resources = [];

                                        foreach ($v as $resource) {
                                            if (\is_array($resource)) {
                                                $resources[] = $resource;
                                            } else {
                                                $resources[] = [
                                                    'resource' => $resource,
                                                ];
                                            }
                                        }

                                        return $resources;
                                    })
                                ->end()
                                ->prototype('array')
                                    ->children()
                                        ->scalarNode('resource')
                                            ->isRequired()
                                            ->cannotBeEmpty()
                                        ->end()
                                        ->enumNode('type')
                                            ->defaultNull()
                                            ->values(['closure', 'container', 'glob', 'php', 'xml', 'yaml', null])
                                        ->end()
                                    ->end()
                                ->end()
                            ->end()
                        ->end()
                    ->end()
                ->end()
            ->end()
        ;
    }

    private function addPingSection(ArrayNodeDefinition $rootNode): void
    {
        $rootNode->children()
            ->arrayNode('ping')
                ->children()
                    ->arrayNode('services')
                        ->arrayPrototype()
                            ->children()
                                ->scalarNode('name')
                                    ->isRequired()
                                    ->cannotBeEmpty()
                                    ->info('The name of the service to ping.')
                                ->end()
                                ->enumNode('type')
                                    ->isRequired()
                                    ->info('The type of the service to be pinged.')
                                    ->values([self::PING_SERVICE_TYPE_DOCTRINE, self::PING_SERVICE_TYPE_PDO])
                                ->end()
                                ->integerNode('interval')
                                    ->defaultValue(20)
                                    ->info('The time (in seconds) between executions of this ping.')
                                    ->min(1)
                                ->end()
                            ->end()
                        ->end()
                    ->end()
                ->end()
            ->end()
        ;
    }
}
